### PR TITLE
Google calendar integration

### PR DESCRIPTION
**Pull request by Mejtika**
## Screenshots (if appropriate)
image

image

image

image

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/DevAdventCalendarCompetition.Services.csproj [hunk 1]
```diff
@@ -16,10 +16,12 @@
     </PackageReference>
     <PackageReference Include="Microsoft.Extensions.Http" Version="3.1.2" />
     <PackageReference Include="Microsoft.VisualStudio.Web.BrowserLink" Version="2.2.0" />
+    <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
```
Comment (WTobor) [inline src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/DevAdventCalendarCompetition.Services.csproj:19]: Do we need both libraries?

File: src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/GoogleCalendarService.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        private readonly HttpClient _httpClient;
+
+        public GoogleCalendarService(HttpClient httpClient)
+        {
+            this._httpClient = httpClient;
+        }
+
+        public async Task<CalendarList> GetAllCalendars()
+        {
+            return await this._httpClient.GetFromJsonAsync<CalendarList>("users/me/calendarList");
```
Comment (WTobor) [inline src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/GoogleCalendarService.cs:23]: please save url to const

File: src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/GoogleCalendarService.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        }
+
+        public async Task<string> CreateNewCalendarWithEvents(string calendarSummary, DateTime startDate, DateTime endDate)
+        {
+            var calendarResponse = await this.CreateCalendar(calendarSummary);
+
+            // For debug
+            calendarResponse.EnsureSuccessStatusCode();
```
Comment (WTobor) [inline src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/GoogleCalendarService.cs:31]: not needed anymore

File: src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/GoogleCalendarService.cs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+            var newCalendar = await calendarResponse.Content.ReadFromJsonAsync<CalendarDto>();
+
+            var eventsResponse = await this.CreateEvents(newCalendar.Id, startDate, endDate);
+            eventsResponse.EnsureSuccessStatusCode();
+
+            if (calendarResponse.IsSuccessStatusCode && eventsResponse.IsSuccessStatusCode)
+            {
+                return $"Pomyślnie utworzono kalendarz: {calendarSummary}";
```
Comment (WTobor) [inline src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/GoogleCalendarService.cs:40]: please return the success response, define the message in the resource file, and use it in the Controller

File: src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/GoogleCalendarService.cs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            }
+
+            return "Wystapił błąd podczas tworzenia nowego kalendarza. Spróbuj ponownie później";
```
Comment (WTobor) [inline src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/GoogleCalendarService.cs:43]: please return the failure response, define the message in the resource file, and use it in the Controller

File: src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/GoogleCalendarService.cs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        }
+
+        private async Task<HttpResponseMessage> CreateCalendar(string calendarSummary)
+        {
+            var newOne = new NewCalendarDto
```
Comment (WTobor) [inline src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/GoogleCalendarService.cs:48]: please rename variable to sth more meaningful ex. newCalendar

File: src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/GoogleCalendarService.cs [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+            {
+                Summary = calendarSummary
+            };
+
+            return await this._httpClient.PostAsJsonAsync("calendars", newOne);
```
Comment (WTobor) [inline src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Services/GoogleCalendarService.cs:53]: please move url to const

**Review by WTobor — COMMENTED** [review]
comments in files :) it looks good!

**Comment by WTobor** [comment]
Have you seen that [nuget](https://developers.google.com/api-client-library/dotnet/apis/calendar/v3)? :)

**Comment by Mejtika** [comment]
Yep, but it stores access token in file or in database. We were talking about that. Just seemed strange at first look.

File: src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Views/GoogleCalendar/_Layout.cshtml [hunk 8]
```diff
@@ -0,0 +1,9 @@
+@{
```
Comment (WTobor) [inline src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Views/GoogleCalendar/_Layout.cshtml:1]: why do we need this?

Comment (Mejtika) [reply src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Views/GoogleCalendar/_Layout.cshtml:1]: I've decoupled the google integration from the Manage controller, so I want the views in GoogleCalendar to use the layout from Manage views and not from Shared

Comment (Mejtika) [reply src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Views/GoogleCalendar/_Layout.cshtml:1]: A better option, in this case, would be to add _ViewStart to the GoogleCalendar Views folder

Comment (Mejtika) [reply src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Views/GoogleCalendar/_Layout.cshtml:1]: But if we want to have only one view for the user, we could get rid of the entire Google Calendar controller and just implement that logic in the Manage controller. One more dependency on service there won't be such a big deal.

Comment (WTobor) [reply src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Views/GoogleCalendar/_Layout.cshtml:1]: @Mejtika @Kamil4921 - do we have any update here?

Comment (Mejtika) [reply src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Views/GoogleCalendar/_Layout.cshtml:1]: @WTobor We're trying

**Comment by Mejtika** [comment]
@WTobor I think it's done. I'm only not sure about which settings from appsettings.development should go to docker and production.

File: src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Controllers/ManageController.cs [hunk 9]
```diff
@@ -272,6 +281,49 @@ public async Task<IActionResult> SetPassword(SetPasswordViewModel model)
...
+        }
+
+        [HttpGet]
+        public async Task<IActionResult> GoogleCalendarIntegration()
+        {
+            var userHasPermissions = await this.CheckIfUserHasPermissions();
+            var model = new GoogleCalendarViewModel
+            {
+                HasPermissions = userHasPermissions,
+                StatusMessage = this.StatusMessage
```
Comment (WTobor) [inline src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Controllers/ManageController.cs:308]: what should be the value of _this.StatusMessage_?

File: src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Views/Manage/GoogleCalendarIntegration.cshtml [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    ViewData.AddActivePage(ManageNavPages.GoogleCalendarIntegration);
+}
+
+<div class="area_title mb-30">
+    <h3>@ViewData["Title"]</h3>
+</div>
+@await Html.PartialAsync("_StatusMessage.cshtml", @Model.StatusMessage)
+@if (!@Model.HasPermissions)
+{
+    <p>W tej sekcji możesz zintegrować DevAdventCalendar z Twoim kalendarzem Google.</p>
```
Comment (WTobor) [inline src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Views/Manage/GoogleCalendarIntegration.cshtml:15]: please use resources

**Comment by WTobor** [comment]
> @WTobor I think it's done. I'm only not sure about which settings from appsettings.development should go to docker and production.

@patrikulus please review PR

File: src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Controllers/ManageController.cs (hunk 9 above)
Comment (Mejtika) [reply src/DevAdventCalendarCompetition/DevAdventCalendarCompetition/Controllers/ManageController.cs:308]: If we're first integrating the value will be
image
If we've been successfully integrated the value of this.StatusMessage will be null, so nothing will be shown to the user

**Review by patrikulus — CHANGES_REQUESTED** [review]
Please add to appsettings.Production.json same parameters as were added to appsettings.Development.json. I think we need to create common appsettings file with those items which aren't unique per environment. I can handle that later, but for the purpose of this PR please just copy them into production settings file.

**Review by patrikulus — APPROVED** [review]

**Pull request merged**